tools/codegen/pkg/manifestmerge/generator.go

```diff
@@ -0,0 +1,562 @@
+				if resultingCRD == nil { // this means we didn't find any file that matched the filter this is ok, we have nothing to do.
+					continue
+				}
+
+				outputFileBaseName := ""
+				if outputFilenamePattern := resultingCRD.GetAnnotations()["api.openshift.io/filename-pattern"]; len(outputFilenamePattern) > 0 {
+					if !strings.Contains(outputFilenamePattern, "MARKERS") {
+						errs = append(errs, fmt.Errorf("crd %q is missing featureset/%q MARKERS from '// +openshift:file-pattern=' %q", crdName, featureSetName, outputFilenamePattern))
+						continue
+					}
+					// TODO this should include clusterProfile once we accommodate it.
+					fileMarker := fmt.Sprintf("-%s", featureSetName)
+					outputFileBaseName = strings.ReplaceAll(outputFilenamePattern, "MARKERS", fileMarker)
+				}
+				if len(outputFileBaseName) == 0 {
+					errs = append(errs, fmt.Errorf("crd %q needs '// +openshift:file-pattern='", crdName))
+					continue
+				}
+				outputFile := filepath.Join(versionPath, outputFileBaseName)
+
+				resultingCRD.SetManagedFields(nil)
+
+				annotations := resultingCRD.GetAnnotations()
+				delete(annotations, "api.openshift.io/filename-pattern")
+				for key := range annotations {
+					if strings.HasPrefix(key, "feature-gate.release.openshift.io/") {
+						delete(annotations, key)
+					}
+				}
+				// TODO probably have cases for different clusterprofiles
+				exception := g.findExceptionForFeatureSet(crdName, featureSetName)
+				if exception != nil && len(exception.ClusterProfilesToInject) > 0 {
+					for _, clusterProfile := range exception.ClusterProfilesToInject {
+						annotations[clusterProfile] = "true"
+					}
+				} else {
+					for _, clusterProfile := range defaultClusterProfilesToInject {
+						annotations[clusterProfile] = "true"
+					}
+				}
+				for key := range annotations {
+					if strings.HasSuffix(key, "-") {
+						toRemove := key[:len(key)-1]
+						delete(annotations, toRemove)
+						delete(annotations, key)
+					}
+				}
+				resultingCRD.SetAnnotations(annotations)
+
+				manifestData, err := kyaml.Marshal(resultingCRD.Object)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("could not encode file %s: %v", outputFile, err))
+					continue
+				}
+
+				if g.verify {
+					existingBytes, err := os.ReadFile(outputFile)
+					if err != nil {
+						errs = append(errs, fmt.Errorf("could not read file %s: %v", outputFile, err))
+						continue
+					}
+					if !bytes.Equal(manifestData, existingBytes) {
+						diff := utils.Diff(existingBytes, manifestData, outputFile)
+
+						return fmt.Errorf("API schema for %s is out of date, please regenerate the API schema:\n%s", outputFile, diff)
+					}
+
+					continue
+				}
+
+				if err := os.WriteFile(outputFile, manifestData, 0644); err != nil {
+					return fmt.Errorf("could not write manifest %s: %w", outputFile, err)
+				}
```

Comment:
> A lot of this is similar to the single iteration case, could this be refactored out to avoid the duplication by passing the exceptions and files in slightly differently?

Exceptions will die in a month and we can refactor.